src/core/server/models/story/story.ts

```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
   return results.filter((section) => section !== null).sort() as string[];
 }
+
+export async function markStoryForArchiving(
+  mongo: Db,
+  tenantID: string,
+  id: string,
+  now: Date
+) {
+  const result = await collection(mongo).findOneAndUpdate(
+    {
+      id,
+      tenantID,
+      $and: [
+        { $or: [{ isArchiving: { $exists: false } }, { isArchiving: false }] },
+        { $or: [{ isArchived: { $exists: false } }, { isArchived: false }] },
+      ],
+    },
+    {
+      $set: {
+        isArchiving: true,
+        closedAt: now,
+        updatedAt: now,
+        archivedAt: now,
+      },
+    },
+    {
+      returnOriginal: false,
+    }
+  );
+
+  return result.value;
+}
+
+export async function markStoryForUnarchiving(
+  mongo: Db,
+  tenantID: string,
+  id: string,
+  now: Date
+) {
+  const result = await collection(mongo).findOneAndUpdate(
+    {
+      id,
+      tenantID,
+      isArchiving: false,
+      isArchived: true,
+    },
+    {
+      $set: {
+        isArchiving: true,
+        closedAt: now,
+        updatedAt: now,
+      },
+    },
+    {
+      returnOriginal: false,
+    }
+  );
+
+  return result.value;
+}
+
+const getCommentCounts = (options: {
+  story: Readonly<Story>;
+  negate: boolean;
+}) => {
+  const {
+    story: { commentCounts },
+    negate,
+  } = options;
+  const multiplier = negate ? -1 : 1;
+
+  const result: RelatedCommentCounts = createEmptyRelatedCommentCounts();
+
+  if (commentCounts.action) {
+    for (const key in commentCounts.action) {
+      if (result.action.hasOwnProperty.call(commentCounts.action, key)) {
+        const value = commentCounts.action[key];
+        result.action[key] = value * multiplier;
+      }
+    }
+  }
+
+  if (commentCounts.moderationQueue) {
+    result.moderationQueue.total =
+      commentCounts.moderationQueue.total * multiplier;
+
+    let key: keyof typeof commentCounts.moderationQueue.queues;
+    for (key in commentCounts.moderationQueue.queues) {
+      if (
+        result.moderationQueue.queues.hasOwnProperty.call(
+          commentCounts.moderationQueue.queues,
+          key
+        )
+      ) {
+        const value = commentCounts.moderationQueue.queues[key];
+        result.moderationQueue.queues[key] = value * multiplier;
+      }
+    }
+  }
+
+  if (commentCounts.status) {
+    let key: keyof typeof commentCounts.status;
+    for (key in commentCounts.status) {
+      if (result.status.hasOwnProperty.call(commentCounts.status, key)) {
+        const value = commentCounts.status[key];
+        result.status[key] = value * multiplier;
+      }
+    }
+  }
+
+  return result;
+};
+
+export async function archiveStory(
+  mongo: MongoContext,
+  redis: AugmentedRedis,
+  tenantID: string,
+  id: string
+) {
+  const targetStory = await collection(mongo.main).findOne({ id, tenantID });
+  if (!targetStory) {
+    throw new StoryNotFoundError(id);
+  }
+
+  if (targetStory.isArchived) {
```

Comment:
Maybe also check if the story is currently already being archived?

Reply:
It will always be archiving cause of the way the `markStoryForArchiving` and `markStoryForUnarchiving` locks the story.